core/src/main/scala/kafka/server/AddPartitionsToTxnManager.scala

```diff
@@ -0,0 +1,170 @@
+
+import scala.collection.mutable
+
+object AddPartitionsToTxnManager {
+  type AppendCallback = Map[TopicPartition, Errors] => Unit
+}
+
+
+class TransactionDataAndCallbacks(val transactionData: AddPartitionsToTxnTransactionCollection,
+                                  val callbacks: mutable.Map[String, AddPartitionsToTxnManager.AppendCallback])
+
+
+class AddPartitionsToTxnManager(config: KafkaConfig, client: NetworkClient, time: Time) 
+  extends InterBrokerSendThread("AddPartitionsToTxnSenderThread-" + config.brokerId, client, config.requestTimeoutMs, time) {
+  
+  private val inflightNodes = mutable.HashSet[Node]()
+  private val nodesToTransactions = mutable.Map[Node, TransactionDataAndCallbacks]()
+  
+  def addTxnData(node: Node, transactionData: AddPartitionsToTxnTransaction, callback: AddPartitionsToTxnManager.AppendCallback): Unit = {
+    nodesToTransactions.synchronized {
+      // Check if we have already (either node or individual transaction). Add the Node if it isn't there.
+      val currentNodeAndTransactionData = nodesToTransactions.getOrElseUpdate(node,
+        new TransactionDataAndCallbacks(
+          new AddPartitionsToTxnTransactionCollection(1),
+          mutable.Map[String, AddPartitionsToTxnManager.AppendCallback]()))
+
+      val currentTransactionData = currentNodeAndTransactionData.transactionData.find(transactionData.transactionalId)
+
+      // Check if we already have txn ID -- this should only happen in epoch bump case. If so, we should return error for old entry and remove from queue.
+      if (currentTransactionData != null) {
+        if (currentTransactionData.producerEpoch() < transactionData.producerEpoch()) {
+          val topicPartitionsToError = mutable.Map[TopicPartition, Errors]()
+          currentTransactionData.topics().forEach { topic =>
+            topic.partitions().forEach { partition =>
+              topicPartitionsToError.put(new TopicPartition(topic.name(), partition), Errors.INVALID_PRODUCER_EPOCH)
+            }
+          }
+          val oldCallback = currentNodeAndTransactionData.callbacks(transactionData.transactionalId())
+          currentNodeAndTransactionData.transactionData.remove(transactionData)
+          oldCallback(topicPartitionsToError.toMap)
+        } else {
+          // We should never see a request on the same epoch since we haven't finished handling the one in queue
+          throw new InvalidRecordException("Received a second request from the same connection without finishing the first.")
+        }
+      }
+      currentNodeAndTransactionData.transactionData.add(transactionData)
+      currentNodeAndTransactionData.callbacks.put(transactionData.transactionalId(), callback)
+      wakeup()
+    }
+  }
+
+  private class AddPartitionsToTxnHandler(node: Node, transactionDataAndCallbacks: TransactionDataAndCallbacks) extends RequestCompletionHandler {
+    override def onComplete(response: ClientResponse): Unit = {
+      inflightNodes.remove(node)
```

Comment:
If we'll have a thread pool of network threads, we'd need to synchronize this too.  So we probably need to add a comment here.

Reply:
I don't recall having a pool of network threads. this would just be the one send thread for now i think.

Reply:
Currently we only have one thread and it might be the case forever, but from this code the threading model is not obvious and it would be useful to have a comment that we don't need synchronization for inflightNodes because inflightNodes is only accessed from methods that are called on the sender's thread.

Reply:
Taking another look -- I see that we access this structure when adding new nodes/data via the request threads, so I think this all should be synchronized after all.